x-pack/plugins/stack_connectors/server/connector_types/inference/inference.ts

```diff
@@ -0,0 +1,225 @@
+   * @param body The stringified request body to be sent in the POST request.
+   */
+  public async performApiRerank({
+    input,
+    query,
+    signal,
+  }: RerankParams & { signal?: AbortSignal }): Promise<RerankResponse> {
+    const response = await this.performInferenceApi(
+      {
+        query,
+        inference_id: this.inferenceId,
+        input,
+        task_type: 'rerank',
+      },
+      false,
+      signal
+    );
+    return response as RerankResponse;
+  }
+
+  /**
+   * responsible for making a POST request to the Inference API sparse embedding task endpoint and returning the response data
+   * @param body The stringified request body to be sent in the POST request.
+   */
+  public async performApiSparseEmbedding({
+    input,
+    signal,
+  }: SparseEmbeddingParams & { signal?: AbortSignal }): Promise<SparseEmbeddingResponse> {
+    const response = await this.performInferenceApi(
+      { inference_id: this.inferenceId, input, task_type: 'sparse_embedding' },
+      false,
+      signal
+    );
+    return response as SparseEmbeddingResponse;
+  }
+
+  /**
+   * responsible for making a POST request to the Inference API text embedding task endpoint and returning the response data
+   * @param body The stringified request body to be sent in the POST request.
+   */
+  public async performApiTextEmbedding({
+    input,
+    inputType,
+    signal,
+  }: TextEmbeddingParams & { signal?: AbortSignal }): Promise<TextEmbeddingResponse> {
+    const response = await this.performInferenceApi(
+      {
+        inference_id: this.inferenceId,
+        input,
+        task_type: 'text_embedding',
+        task_settings: {
+          input_type: inputType,
+        },
+      },
+      false,
+      signal
+    );
+    return response as TextEmbeddingResponse;
+  }
+
+  /**
+   * responsible for making a POST request to the Inference API chat completetion task endpoint and returning the response data
+   * @param body The stringified request body to be sent in the POST request.
+   */
+  private async performInferenceApi(
+    params: InferenceInferenceRequest,
+    asStream: boolean = false,
+    signal?: AbortSignal
+  ): Promise<InferenceInferenceResponse> {
+    try {
+      const response = await this.esClient?.inference.inference(params, { asStream, signal });
+      this.logger.info(
+        `Perform Inference endpoint for task type "${this.taskType}" and inference id ${this.inferenceId}`
+      );
+      // TODO: const usageMetadata = response?.data?.usageMetadata;
+      return response;
+    } catch (err) {
+      this.logger.error(`error perform inference endpoint API: ${err}`);
+      throw err;
+    }
+  }
+
+  private async streamAPI({
+    input,
+    signal,
+  }: ChatCompleteParams & { signal?: AbortSignal }): Promise<StreamingResponse> {
+    const response = await this.performInferenceApi(
+      { inference_id: this.inferenceId, input, task_type: this.taskType as InferenceTaskType },
```

Comment:
Should `streamAPI` take an optional `taskType` argument? It's confusing to me that `performApiCompletion` sets `task_type: completion` but its inferred from the connector for `performApiCompletionStream`
```suggestion
      { inference_id: this.inferenceId, input, task_type: taskType ?? this.taskType as InferenceTaskType },
```

Reply:
TaskType is single per connector, once you've created connector you cannot change the task type, because connector instance correspond to the Inference Endpoint, which has a configuration set up per taskType and service provider configuration. For example you cannot use the same OpenAI connector for completion task and text_embedding, because it requires a different setup for Inference Endpoint. It is always service + taskType

Reply:
It's confusing that the other tasks here have a defined `taskType` and streaming is the only one that leaves it up to the connector